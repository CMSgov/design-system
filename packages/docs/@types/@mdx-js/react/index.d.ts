declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
