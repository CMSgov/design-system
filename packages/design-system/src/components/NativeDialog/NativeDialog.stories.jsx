import React, { Component, useState } from 'react';
import { NativeDialog as DialogComponent } from './NativeDialog';

export default {
  title: 'Components/Native Dialog',
  component: DialogComponent,
  args: {},
  parameters: {
    docs: {
      source: {
        type: 'code',
      },
    },
  },
};

const Template = ({ data, ...args }) => <DialogComponent {...args} />;

const Modal = Template.bind({});

const Drawer = Template.bind({});
