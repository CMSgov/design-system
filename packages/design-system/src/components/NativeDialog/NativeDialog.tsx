import React from 'react';
import { useRef, useEffect, useLayoutEffect, DialogHTMLAttributes } from 'react';
import dialogPolyfill from './polyfill';

export interface NativeDialogProps extends Omit<DialogHTMLAttributes<HTMLElement>, 'children'> {
  children: React.ReactNode;
  /**
   * Pass `true` to have the dialog close when its backdrop pseudo-element is clicked
   */
  backdropClickExits?: boolean;
  /**
   * Function called to close dialog.
   */
  exit: (...args: any[]) => any;
  /**
   * A custom `id` attribute for the dialog element
   */
  id?: string;
  /**
   * Determines which native dialog method is called to open the dialog. If the dialog
   * is opened with `showModal`, focus will be trapped inside the dialog, and it will
   * need to be closed before the rest of the page can be interacted with. See the
   * [`showModal` MDN documentation](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal)
   * for more details.
   */
  showModal?: boolean;
}

export const NativeDialog = ({
  children,
  exit,
  showModal,
  backdropClickExits,
  ...dialogProps
}: NativeDialogProps) => {
  const dialogRef = useRef(null);

  // Register dialog with the polyfill if necessary
  useLayoutEffect(() => {
    // The registerDialog function itself determines if the polyfill needs to be applied
    dialogPolyfill.registerDialog(dialogRef.current);
  });

  // Open and close the dialog with the appropriate method on mount and unmount
  useEffect(() => {
    const dialogNode = dialogRef.current;
    showModal ? dialogNode.showModal() : dialogNode.show();
    return () => {
      dialogNode.close();
    };
  }, [showModal]);

  // Bind and unbind event listeners on mount and unmount
  useEffect(() => {
    const dialogNode = dialogRef.current;

    const handleCancel = (event) => {
      event.preventDefault();
      exit();
    };
    dialogNode.addEventListener('cancel', handleCancel);

    const handleClick = (event) => {
      const rect = dialogNode.getBoundingClientRect();
      const isInDialog =
        rect.top <= event.clientY &&
        event.clientY <= rect.top + rect.height &&
        rect.left <= event.clientX &&
        event.clientX <= rect.left + rect.width;
      if (!isInDialog) {
        exit();
      }
    };
    if (backdropClickExits) {
      dialogNode.addEventListener('click', handleClick);
    }

    return () => {
      dialogNode.removeEventListener('cancel', handleCancel);
      if (backdropClickExits) {
        dialogNode.removeEventListener('click', handleClick);
      }
    };
  }, [exit, backdropClickExits]);

  return (
    <dialog ref={dialogRef} {...dialogProps}>
      {children}
    </dialog>
  );
};

export default NativeDialog;
