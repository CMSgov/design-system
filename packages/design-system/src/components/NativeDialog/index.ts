export * from './NativeDialog';
export * from './shim';
