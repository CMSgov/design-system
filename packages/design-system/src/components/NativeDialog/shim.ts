/**
 * This polyfill originally came from https://github.com/GoogleChrome/dialog-polyfill/,
 * but we've stripped it down to only what we need to not crash node environments that
 * do not yet support the dialog element.
 */

/**
 * Add shims for dialog functions and properties to the provided element object.
 */
function addDialogProperties(dialog: any) {
  function handleKey(event) {
    if (event.keyCode === 27) {
      event.preventDefault();
      event.stopPropagation();
      dialog.close();
    }
  }

  function setOpen(open: boolean) {
    if (open) {
      dialog.hasAttribute('open') || dialog.setAttribute('open', '');
      document.addEventListener('keydown', handleKey);
    } else {
      dialog.removeAttribute('open');
      document.removeEventListener('keydown', handleKey);
    }
  }

  dialog.show = () => setOpen(true);
  dialog.showModal = () => setOpen(true);
  dialog.close = () => {
    setOpen(false);
    const cancelEvent = new CustomEvent('close', {
      bubbles: false,
      cancelable: true,
    });
    dialog.dispatchEvent(cancelEvent);
  };

  // Set a11y role. Browsers that support dialog implicitly know this already.
  if (!dialog.hasAttribute('role')) {
    dialog.setAttribute('role', 'dialog');
  }

  if (!('returnValue' in dialog)) {
    dialog.returnValue = '';
  }

  Object.defineProperty(dialog, 'open', {
    set: setOpen,
    get: dialog.hasAttribute.bind(dialog, 'open'),
  });
}

export function shimDialogElement(element: HTMLDialogElement) {
  if (!element.showModal) {
    if (element.localName !== 'dialog') {
      throw new Error('Failed to polyfill dialog: The element is not a dialog.');
    }
    addDialogProperties(element);
  }
}
