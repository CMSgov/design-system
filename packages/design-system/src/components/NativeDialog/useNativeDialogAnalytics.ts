import { useRef, useEffect } from 'react';
import getAnalyticsContentFromRefs from '../analytics/getAnalyticsContentFromRefs';
import usePrevious from '../utilities/usePrevious';

export interface UseNativeDialogAnalyticsProps {
  isOpen: boolean;
  onOpen: (content?: string) => any;
  onClose?: (content?: string) => any;
}

/**
 * This hook is for any component based on `NativeDialog` that needs analytics that fire
 * upon opening and closing. You provide it the `isOpen` prop being passed to the
 * `NativeDialog` and handlers for `onOpen` and `onClose`, which will be called with the
 * rendered heading content based on where you apply the ref this hook returns. Example:
 *
 * const headingRef = useNativeDialogAnalytics({
 *   isOpen: props.isOpen,
 *   onOpen: (content) => {
 *     // Send an event
 *   },
 *  onClose: (content) => {
 *     // Send an event
 *   }
 * })
 *
 * return (
 *   <NativeDialog {...props}>
 *     <h1 ref={headingRef}>Hello World</h1>
 *   </NativeDialog>
 * )
 */
export function useNativeDialogAnalytics({
  isOpen,
  onOpen,
  onClose,
}: UseNativeDialogAnalyticsProps) {
  const headingRef = useRef();
  const prevIsOpen = usePrevious(isOpen);
  useEffect(() => {
    const headingContent = getAnalyticsContentFromRefs([headingRef]);
    // We want to call onOpen when it's open once and only once.
    // We want to call onClose only when it was open and then closed.
    if (isOpen && isOpen !== prevIsOpen) {
      // It either started open or was closed and now is open
      onOpen(headingContent);
    } else if (!isOpen && prevIsOpen === true) {
      // Only trigger if it was previously open
      onClose(headingContent);
    }
  }, [isOpen, prevIsOpen]);

  return headingRef;
}
