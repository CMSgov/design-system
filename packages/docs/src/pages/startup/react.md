---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
