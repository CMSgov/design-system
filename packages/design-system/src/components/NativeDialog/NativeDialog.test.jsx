import NativeDialog from './NativeDialog';
import React from 'react';
import { render, screen } from '@testing-library/react';

const defaultProps = {
  children: <p>Foo</p>,
};

function renderNativeDialog(props = {}) {
  // eslint-disable-next-line react/no-children-prop
  return render(<NativeDialog data-testid="test-dialog" {...defaultProps} {...props} />);
}

describe('NativeDialog', function () {
  it('renders with custom className', () => {
    renderNativeDialog({
      className: 'custom-class',
    });
    expect(screen.getByTestId('test-dialog')).toMatchSnapshot();
  });
});
