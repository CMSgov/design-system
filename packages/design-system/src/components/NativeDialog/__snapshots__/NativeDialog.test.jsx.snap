// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NativeDialog renders with custom className 1`] = `
<dialog
  class="custom-class"
  data-testid="test-dialog"
>
  <p>
    Foo
  </p>
</dialog>
`;
